### ISSUE TITLE
How to make computations deterministic 

### ISSUE THREAD

**Issue by GarrettLee** [NVIDIA/apex#454]
When using PyTorch, I can set the following options to make computations completely reproducible:

```python
torch.manual_seed(0)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
```

However, if I wrap my model with apex, such deterministic computation is no longer guaranteed. Are there any option I can set to make it possible? Thanks!

**Comment by ptrblck** [comment]
Hi @GarrettLee,
as explained in the [Reproducibility docs](https://pytorch.org/docs/stable/notes/randomness.html) there are some PyTorch functions using CUDA which are non-deterministic:
> There are some PyTorch functions that use CUDA functions that can be a source of non-determinism. One class of such CUDA functions are atomic operations, in particular atomicAdd, where the order of parallel additions to the same value is undetermined and, for floating-point variables, a source of variance in the result. PyTorch functions that use atomicAdd in the forward include torch.Tensor.index_add_(), torch.Tensor.scatter_add_(), torch.bincount().

> A number of operations have backwards that use atomicAdd, in particular torch.nn.functional.embedding_bag(), torch.nn.functional.ctc_loss() and many forms of pooling, padding, and sampling. There currently is no simple way of avoiding non-determinism in these functions.

Are you getting exactly reproducible results without `amp`?
If so, could you post a (minimal) code snippet so that we can have a look?

**Comment by ashok-arjun** [comment]
@ptrblck Has this issue been solved?

Despite following the reproducibility guidelines, I am not getting the same results across multiple runs with `amp` although I am getting the same results across runs without `amp`.

Thanks!
